src/shared/modules/react-native-locale/ReactNativeLocaleProvider.js

```diff
@@ -0,0 +1,96 @@
+import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
+import PropTypes from 'prop-types';
+import { RawIntlProvider, createIntl, createIntlCache } from 'react-intl';
+import * as localize from 'react-native-localize';
+import ReactNativeLocaleContext from './context';
+
+const Provider = ReactNativeLocaleContext.Provider;
+
+const findLocaleById = (locales, localeId) => locales.find((locale) => locale.id === localeId);
+const getInitialLocale = (locales, defaultLocale, initialLocaleId) => {
+    if (locales.length === 0) {
+        throw new Error('Locales array is empty. Please provide at least one locale (default).');
+    }
+
+    if (initialLocaleId) {
+        const locale = findLocaleById(locales, initialLocaleId);
+
+        if (!locale) {
+            throw new Error(`Unknown initial locale id: ${initialLocaleId}.`);
+        }
+
+        return locale;
+    }
+
+    const localeId = localize.findBestAvailableLanguage(locales.map((locale) => locale.id))?.languageTag;
+
+    return findLocaleById(locales, localeId) ?? defaultLocale;
+};
+
+const ReactNativeLocaleProvider = ({
+    children,
+    locales,
+    initialLocaleId,
+    onLocaleChange,
+}) => {
+    const intlCache = useRef(createIntlCache());
+    const [currentLocale, setCurrentLocale] = useState(getInitialLocale(locales, locales[0], initialLocaleId));
+
+    const handleLocaleChange = useCallback((localeId) => {
+        const nextLocale = findLocaleById(locales, localeId);
+
+        if (!nextLocale) {
+            return;
+        }
+
+        setCurrentLocale(nextLocale);
+
+        if (localeId !== currentLocale.id) {
+            onLocaleChange?.(nextLocale.id);
```

Comment:
We will maintain this? 😛

Reply:
Well, no harm's done if we keep the callback as optional. I'd rather offer the possibility of attaching a callback now than later realize we need it for some reason and not having it in place.

Reply:
I do not like having a provider cause other unexpected effects from a callback sent to him. To me whoever consumes it should receive the `locale` and a `setLocale` and when the locale changes or whenever the developer wants to modify the locale he should do it in the `useCallback` where he calls `setLocale`. But the rest can give their opinion

Reply:
> I do not like having a provider cause other unexpected effects from a callback sent to him. To me whoever consumes it should receive the `locale` and a `setLocale` and when the locale changes or whenever the developer wants to modify the locale he should do it in the `useCallback` where he calls `setLocale`.

That's exactly what I had in mind. If the possibility to change the language in-app is offered by any given app, the approach one should take is the one you've mentioned. However, that only works if we're considering only React components listen to locale changes. What if we come across a scenario where a module is interested in locale changes as well?

Reply:
Then you would just need to use create a hook with a `useEffect` with the `locale` as a dependency and import the module and have it return whatever the module returns 😛 less magic, more coherent in my opinion.
@satazor @fsdiogo what do you think?

Reply:
> Then you would just need to create a hook with a `useEffect` with the `locale` as a dependency and import the module and have it return whatever the module returns 😛 less magic, more coherent in my opinion.
> @satazor @fsdiogo what do you think?

If the module is not a React component nor does have dependencies on React, it really doesn't make much sense to create a React component with a `useEffect` just for that purpose. Also, that component would not render anything at all which makes it kind of useless. 😕